### PR TITLE
Added Unit tests and Integration Tests for all OSes

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py src/test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -118,8 +121,17 @@ pipeline {
                     }
                     post { always { deleteDirWin() } }
                 }
-                stage('Unit') {
-                    agent any
+                stage('Linux Unit') {
+                    agent { label 'linux' }
+                    steps {
+                        unstash 'StanSetup'
+                        setupCXX(false)
```
Comment (seantalts) [inline Jenkinsfile]: By default the linux nodes have clang as their compiler I believe; we'll need to set it specifically here.

File: Jenkinsfile [hunk 2]
```diff
@@ -104,40 +105,77 @@ pipeline {
...
-                unstash 'StanSetup'
-                setupCXX()
-                runTests("src/test/integration", separateMakeStep=false)
+            parallel {
+                stage('Integration Windows') {
+                    agent { label 'windows' }
+                    steps {
+                        deleteDirWin()
+                            unstash 'StanSetup'
+                            setupCXX(true,"${env.CXX}")
```
Comment (seantalts) [inline Jenkinsfile]: this can just be env.CXX, doesn't need to be quoted into a string.

File: Jenkinsfile [hunk 3]
```diff
@@ -3,12 +3,11 @@ import org.stan.Utils
 
 def utils = new org.stan.Utils()
 
-def setupCXX(failOnError = true) {
+def setupCXX(failOnError = true, CXX = env.GCC) {
```
Comment (seantalts) [inline Jenkinsfile]: I think it makes more sense to have the default be env.CXX because it's used more that way

**Comment by serban-nicusor-toptal** [comment]
!!! This is just a test, please ignore !!!

('stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan', 1.01)
('stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan', 0.99)
('stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan', 0.98)
('stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan', 0.98)
('stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan', 0.99)
('stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan', 1.0)
('compilation', 1.02)
('stat_comp_benchmarks/benchmarks/arK/arK.stan', 0.99)
('stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan', 1.0)
('stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan', 1.0)
('stat_comp_benchmarks/benchmarks/sir/sir.stan', 1.0)
('stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan', 0.77)
('stat_comp_benchmarks/benchmarks/garch/garch.stan', 0.99)
('stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan', 1.02)
('stat_comp_benchmarks/benchmarks/arma/arma.stan', 0.99)
Result: 0.98242302359
Commit hash: d52531c66e40911a42dc7613fc6a1cf36fc26a13

**Comment by seantalts** [comment]
Output looks great to me.

File: Jenkinsfile [hunk 4]
```diff
@@ -66,7 +67,7 @@ pipeline {
             }
         }
         stage('Linting & Doc checks') {
-            agent any
+            agent {label 'linux'}
```
Comment (seantalts) [inline Jenkinsfile]: does this have to be linux now?

File: Jenkinsfile (hunk 1 above)
Comment (seantalts) [reply Jenkinsfile]: I think you have the correct line commented out here, possibly? Or is this still waiting for GCC fixes? If so can you put the issue # in a comment directly into the Jenkinsfile for us?

File: Jenkinsfile [hunk 5]
```diff
@@ -129,15 +141,42 @@ pipeline {
...
+            parallel {
+                stage('Integration Windows') {
+                    agent { label 'windows' }
+                    steps {
+                        deleteDirWin()
+                            unstash 'StanSetup'
+                            setupCXX()
+                            bat "make -j${env.PARALLEL} test-headers"
+                            setupCXX(false)
+                            runTestsWin("src/test/integration")
```
Comment (seantalts) [inline Jenkinsfile]: I think this is right. But I also had the impression a separate make step adds a bunch of time on Windows - is that true? If it adds more than a few minutes we could just change this to separateMakeStep=false to save time.

File: Jenkinsfile (hunk 4 above)
Comment (serban-nicusor-toptal) [reply Jenkinsfile]: I think this speeds up the build a bit since windows tend to be a bit slower and have a longer spawn time.

**Comment by seantalts** [comment]
`agent any` wouldn't use windows nodes, it would use either Linux or Mac.
We have the Windows node set up to only take jobs that are specifically
asking for windows.

File: Jenkinsfile (hunk 1 above)
Comment (serban-nicusor-toptal) [reply Jenkinsfile]: If I switch it to gcc it will not build anymore.

**Comment by serban-nicusor-toptal** [comment]
I see, my bad then. Will revert to any

File: Jenkinsfile (hunk 5 above)
Comment (serban-nicusor-toptal) [reply Jenkinsfile]: This takes 5 minutes, check it out here: https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-2761/10/pipeline

**Comment by serban-nicusor-toptal** [comment]
- - - - - - - - - - - - - - - - - - - - -\r\n| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) |\r\n| ------------- |------------- | ------------- | ------------- | ------------- |\r\n| gp_pois_regr | 4.03 | 3.99 | 1.01 | 0.97% faster | \r\n| low_dim_corr_gauss | 0.02 | 0.02 | 0.99 | -0.55% slower | \r\n| eight_schools | 0.1 | 0.09 | 1.02 | 2.22% faster | \r\n| gp_regr | 0.04 | 0.04 | 0.97 | -2.59% slower | \r\n| irt_2pl | 6.68 | 6.67 | 1.0 | 0.2% faster | \r\n| low_dim_gauss_mix_collapse | 10.69 | 10.69 | 1.0 | 0.02% faster | \r\n| pkpd | 0.37 | 0.37 | 1.0 | 0.05% faster | \r\n| sir | 122.38 | 117.33 | 1.04 | 4.12% faster | \r\n| low_dim_gauss_mix | 3.26 | 3.26 | 1.0 | -0.04% slower | \r\n| arK | 2.37 | 2.36 | 1.01 | 0.68% faster | \r\n| arma | 0.83 | 0.83 | 1.0 | 0.17% faster | \r\n| garch | 0.63 | 0.62 | 1.0 | 0.37% faster | \r\n| compilation | 88.92 | 87.59 | 1.02 | 1.49% faster | \r\n- - - - - - - - - - - - - - - - - - - - -\r\nMean result: 1.00955339885\r\nCommit hash: 4dc1d64cbe31acba391946cde9f45025a2c91c23\r\n[Jenkins Console Log](https://jenkins.mc-stan.org/job/stan/view/change-requests/job/PR-2761/134/consoleFull)\r\n[Blue Ocean](https://jenkins.mc-stan.org/blue/organizations/jenkins/stan/detail/PR-2761/134/pipeline)\r\n- - - - - - - - - - - - - - - - - - - - -\r\n<details><summary>Machine information</summary>\r\nOperating System: \r\nProductName:	Mac OS X
ProductVersion:	10.11.6
BuildVersion:	15G22010\r\n\r\nCPU: \r\nIntel(R) Xeon(R) CPU E5-1680 v2 @ 3.00GHz\r\n\r\nG++: \r\nConfigured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.6.0
Thread model: posix\r\n\r\nClang: \r\nApple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.6.0
Thread model: posix\r\n\r\n</details>\r\n

**Comment by serban-nicusor-toptal** [comment]
```Hello world!```

**Comment by serban-nicusor-toptal** [comment]
- - - - - - - - - - - - - - - - - - - - -\r\n| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) |\r\n| ------------- |------------- | ------------- | ------------- | ------------- |\r\n| gp_pois_regr | 4.93 | 5.09 | 0.97 | -3.25% slower | \r\n| low_dim_corr_gauss | 0.02 | 0.02 | 0.99 | -1.25% slower | \r\n| eight_schools | 0.14 | 0.15 | 0.97 | -3.44% slower | \r\n| gp_regr | 0.06 | 0.06 | 1.03 | 3.3% faster | \r\n| irt_2pl | 10.14 | 10.35 | 0.98 | -2.07% slower | \r\n| low_dim_gauss_mix_collapse | 14.15 | 13.78 | 1.03 | 2.62% faster | \r\n| pkpd | 0.5 | 0.52 | 0.96 | -4.02% slower | \r\n| sir | 193.31 | 197.41 | 0.98 | -2.12% slower | \r\n| garch | 1.08 | 1.09 | 0.99 | -1.16% slower | \r\n| arK | 2.93 | 2.93 | 1.0 | -0.01% slower | \r\n| arma | 1.26 | 1.26 | 1.0 | -0.1% slower | \r\n| low_dim_gauss_mix | 4.48 | 4.48 | 1.0 | -0.11% slower | \r\n| compilation | 164.82 | 156.93 | 1.05 | 4.79% faster | \r\n- - - - - - - - - - - - - - - - - - - - -\r\nMean result: 0.99262870965\r\nCommit hash: 4dc1d64cbe31acba391946cde9f45025a2c91c23\r\n[Jenkins Console Log](https://jenkins.mc-stan.org/job/stan/view/change-requests/job/PR-2761/136/consoleFull)\r\n[Blue Ocean](https://jenkins.mc-stan.org/blue/organizations/jenkins/stan/detail/PR-2761/136/pipeline)\r\n- - - - - - - - - - - - - - - - - - - - -\r\n<details><summary>Machine information</summary>\r\nOperating System: \r\nProductName:	Mac OS X\nProductVersion:	10.13.4\nBuildVersion:	17E199\r\n\r\nCPU: \r\nIntel(R) Core(TM) i7 CPU         870  @ 2.93GHz\r\n\r\nG++: \r\nConfigured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1\nApple LLVM version 9.1.0 (clang-902.0.39.2)\nTarget: x86_64-apple-darwin17.5.0\nThread model: posix\nInstalledDir: /Library/Developer/CommandLineTools/usr/bin\r\n\r\nClang: \r\nApple LLVM version 9.1.0 (clang-902.0.39.2)\nTarget: x86_64-apple-darwin17.5.0\nThread model: posix\nInstalledDir: /Library/Developer/CommandLineTools/usr/bin\r\n\r\n</details>\r\n

**Comment by rok-cesnovar** [comment]
g++ tests now work and so should Windows so let's see if a merge of recent develop fixes this now.

File: makefile [hunk 6]
```diff
@@ -122,12 +122,12 @@ clean-dox:
 
 clean-deps:
 	@echo '  removing dependency files'
-	$(RM) $(call findfiles,.,*.d)
+	$(RM) $(call findfiles,./,*.d)
```
Comment (rok-cesnovar) [inline makefile:125]: This is a minor cleanup I could also do as part of a separate PR if you feel.

File: runTests.py [hunk 7]
```diff
@@ -142,7 +142,8 @@ def main():
 
     for batch in batched(tests):
         modelHpp = modelDependencies(batch)
-        makeTest(" ".join(modelHpp), inputs.j)
+        if len(modelHpp) > 0:
+            makeTest(" ".join(modelHpp), inputs.j)
```
Comment (rok-cesnovar) [inline runTests.py:146]: This is also minor cleanup to not run this make step if no model was needed in this batch of tests.

File: Jenkinsfile [hunk 8]
```diff
@@ -216,18 +225,31 @@ pipeline {
             }
         }
         stage('Integration') {
+            parallel {
+                stage('Integration Linux') {
```
Comment (rok-cesnovar) [inline Jenkinsfile:229]: Windows will be added here once we fix integration tests there.

File: .travis.yml [hunk 9]
```diff
@@ -69,52 +61,6 @@ matrix:
         - MATRIX_EVAL="CXX=clang++-6.0 && CXXFLAGS_OS=-stdlib=libc++"
           TESTFOLDER="src/test/unit/services src/test/unit/variational src/test/unit/version_test.cpp"
           PARALLEL=2
-    - <<: *linux_gcc
```
Comment (rok-cesnovar) [inline .travis.yml:72]: No need to duplicate these tests anymore on Travis. I left clang ubuntu tests which should stay in my opinion. 

Linux is the only platform where two different compilers are used commonly. On windows clang++ does not currently work with Stan, and g++ on macOS is very rare.

We should maybe switch to 18.04, this is still using 14.04. But that is a separate issue.

File: src/test/unit/lang/stanc_helper_test.cpp [hunk 10]
```diff
@@ -28,14 +28,14 @@ TEST(commandStancHelper, printStancHelp) {
 
 bool create_test_file(const std::string& path, const std::string& program) {
   std::string cmd_setup = "mkdir -p test/test-models";
-  system(cmd_setup.c_str());
+  int return_code = system(cmd_setup.c_str());
```
Comment (rok-cesnovar) [inline src/test/unit/lang/stanc_helper_test.cpp:31]: This was added as g++ threw a warning that the return value is not used. And warnings are treated as errors.

**Comment by rok-cesnovar** [comment]
This is now ready for review and passing tests. I had to add an additional fix to a test that was caught by g++ on Linux.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal would you mind double-checking that Jenkins/Travis changes are valid? I asked @t4c1 to look at the rest to get this moving along.

File: src/test/unit/lang/stanc_helper_test.cpp (hunk 10 above)
Comment (t4c1) [reply src/test/unit/lang/stanc_helper_test.cpp:31]: I think the usual solution if you want to ignore a return value is to cast it to void. I think that should also avoid the warning.

File: src/test/unit/lang/stanc_helper_test.cpp [hunk 11]
```diff
@@ -28,14 +28,14 @@ TEST(commandStancHelper, printStancHelp) {
...
   std::string cmd = "echo ";
   cmd += "\"";
   cmd += program;
   cmd += "\"";
   cmd += " > ";
   cmd += path;
-  int return_code = system(cmd.c_str());
+  return_code = system(cmd.c_str());
```
Comment (t4c1) [inline src/test/unit/lang/stanc_helper_test.cpp:38]: If you dont want to actually ignore errors in the first call:
```suggestion
  return_code |= system(cmd.c_str());
```

Comment (rok-cesnovar) [reply src/test/unit/lang/stanc_helper_test.cpp:38]: I will just cat to void. The return code was not used before, so that is the least invasive change.

Comment (rok-cesnovar) [reply src/test/unit/lang/stanc_helper_test.cpp]: Done.

**Comment by serban-nicusor-toptal** [comment]
Jenkins/Travis code looks good to me, it seems to [fail](https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-2761/40/pipeline) in Jenkins during a test in the last build. I hope the old imac we have can handle the build fine, else we can change the `osx` label to `gelman-group-mac`.

**Comment by rok-cesnovar** [comment]
Yes, the cast to void did not work so I went with Tadej's other suggestion. But I am unable to push it as github is having some issue atm. https://www.githubstatus.com/

**Comment by rok-cesnovar** [comment]
Thanks! I think the old-imac should hande it, but lets see.

**Comment by serban-nicusor-toptal** [comment]
I had trouble posting here too, should be back in a few minutes.
